### PR TITLE
Adding compatibility testing to ensure the update from current master is possible

### PR DESCRIPTION
**Pull request by bdematte**
Adding test case where boilerplate has been initialised with the former version, to ensure smooth migration and no issue while deploying the new version.

### CODE REVIEW

File: test/case/03_test_compatibility [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash -x
```
Comment (2uasimojo) [inline test/case/03_test_compatibility:1]: Can we name this file something like `03_test_update_from_master`?

File: test/case/03_test_compatibility [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+if [ $? -ne 0 ] ; then
+	exit $?
+fi
+
+add_convention test/test_base_convention
+make update_boilerplate
+
+check_update
+
```
Comment (2uasimojo) [inline test/case/03_test_compatibility:34]: Did you mean to `reset_boilerplate_version`, `make update_boilerplate`, and `check_update` again here? Else all we've done is checked `master`, which we know already passed or it wouldn't have gotten merged :P

File: test/lib.sh [hunk 3]
```diff
@@ -26,7 +26,7 @@ add_cleanup() {
 }
 
 empty_repo() {
-    tmpd=$(mktemp -d)
+    tmpd=$(mktemp -d -t boilerplate_test_XXXXXXXX)
```
Comment (2uasimojo) [inline test/lib.sh:29]: micronit: any reason underscores here and hyphens [below](https://github.com/openshift/boilerplate/pull/17/files/a45b4cf538a8ad3837bbaff5b5f04b1264d3641b..b6d0ef4e5e6853ce4322b9f111268f688a0abd8c#diff-e1b2b21859e872e9af54745ed0a1b5c4R163)?

File: test/lib.sh [hunk 4]
```diff
@@ -154,3 +154,34 @@ handle_error_counter() {
 	
 	exit 0
 }
+
+## new_boilerplate_repo POS
+#
+# Make a new clone of boilerplate, checking out POS (may it be branch or commit ID)
+# :param POS: The position in the git repository to checkout (branch or commit ID)
+new_boilerplate_repo() {
+	pushd $(mktemp -d -t boilerplate-clone-XXXXXXXX) > /dev/null
```
Comment (2uasimojo) [inline test/lib.sh:163]: `add_cleanup` this dir

File: test/case/03_test_compatibility (hunk 2 above)
Comment (bdematte) [reply test/case/03_test_compatibility:34]: yes, needs to finish writing the test case

File: test/lib.sh (hunk 3 above)
Comment (bdematte) [reply test/lib.sh:29]: nope, fixed

File: test/case/03_test_update_from_master [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+make update_boilerplate
+
+check_update 
+
+if [ $? -ne 0 ] ; then
+	exit $?
+fi
+
+add_convention test/test_base_convention
```
Comment (bdematte) [inline test/case/03_test_update_from_master:30]: For now, it is failing, pending https://github.com/openshift/boilerplate/pull/16 to be merged

**Review by 2uasimojo — APPROVED** [review]
Only a couple of these comments are especially important, foremost of which is the bug in `new_boilerplate_repo`. But since the [breaking line](https://github.com/openshift/boilerplate/pull/17/files#diff-e1b2b21859e872e9af54745ed0a1b5c4R146) is currently not reached by tests, it's not a hard blocker for this PR.

Since I've got a broad refactor to do today that's going to touch everything, I'm going to merge this as is, and we can come back and address these issues in a subsequent PR.

/lgtm
/approve

File: test/lib.sh [hunk 6]
```diff
@@ -133,3 +133,34 @@ add_convention() {
         echo $2 >> $1/boilerplate/update.cfg
     fi
 }
+
+## new_boilerplate_repo POS
+#
+# Make a new clone of boilerplate, checking out POS (may it be branch or commit ID)
+# :param POS: The position in the git repository to checkout (branch or commit ID)
+new_boilerplate_repo() {
+	pushd $(mktemp -d -t boilerplate_clone_XXXXXXXX) > /dev/null
```
Comment (2uasimojo) [inline test/lib.sh:142]: Need to `add_cleanup`, either here or at the consumer.

Comment (2uasimojo) [reply test/lib.sh:142]: Done.

File: test/lib.sh [hunk 7]
```diff
@@ -133,3 +133,34 @@ add_convention() {
...
+	pwd
+	git clone https://github.com/openshift/boilerplate.git > /dev/null
+	if [ $# = 1 ] ; then
+		git checkout $1
+	fi
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply test/lib.sh:147]: Done.

File: test/lib.sh [hunk 8]
```diff
@@ -133,3 +133,34 @@ add_convention() {
...
     fi
 }
+
+## new_boilerplate_repo POS
+#
+# Make a new clone of boilerplate, checking out POS (may it be branch or commit ID)
+# :param POS: The position in the git repository to checkout (branch or commit ID)
+new_boilerplate_repo() {
+	pushd $(mktemp -d -t boilerplate_clone_XXXXXXXX) > /dev/null
+	pwd
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply test/lib.sh:143]: Done. (Turns out `empty_repo` was the only other function that needed this.)

File: test/lib.sh [hunk 9]
```diff
@@ -133,3 +133,34 @@ add_convention() {
...
+		git checkout $1
+	fi
+	popd > /dev/null
+}
+
+## override_boilerplate_version NEW_PATH
+#
+# Override the boilerplate version to be used for the testing. 
+# :param NEW_PATH: A clone of boilerplate to be used for the future steps
+override_boilerplate_version() {
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply test/lib.sh:155]: Done

File: test/lib.sh [hunk 10]
```diff
@@ -133,3 +133,34 @@ add_convention() {
...
+	if [ -d $1 ] ; then
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply test/lib.sh:156]: Done

File: test/lib.sh [hunk 11]
```diff
@@ -133,3 +133,34 @@ add_convention() {
...
+		BOILERPLATE_GIT_REPO=$1
+	fi
+}
+
+## reset_boilerplate_version 
+#
+# Reset the boilerplate version to be used to the 'tested' clone. 
+reset_boilerplate_version() {
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply test/lib.sh:164]: Done.

**Review by 2uasimojo — COMMENTED** [review]
Addressed all these comments via #30.

**Pull request merged**